src/extra/components/ppu.py

```diff
@@ -155,22 +155,29 @@ def train_ids(
         return train_ids
 
     def trains(
-        self, split_sequence: bool = False, offset: int = 0
+        self,
+        data: Union[DataCollection, SourceData, KeyData] = None,
+        *,
+        split_sequence: bool = False,
+        offset: int = 0,
     ) -> Union[DataCollection, List[DataCollection]]:
         """Returns a subset of the data only with Trains selected by the PPU.
 
         Args:
+            data: Data to filter. If set to None (defaut) use the data used at initialization.
             split_sequence (bool, optional): Split data per PPU trigger sequence. Defaults to False.
             offset (int, optional): offset to apply to train IDs to be selected. Defaults to 0.
 
         Returns:
             Union[DataCollection, List[DataCollection]]:
                 DataCollection(s) containing only trains triggered by the PPU
         """
+        data = data or self.data
```

Comment:
Good to make this generic. Strictly speaking, the EXtra component is thus not a PPU component any more, but a "trainSelector" component. I wonder if there should be such class per se. I see that this addition to the ``trains()`` method makes the PPU more flexible, but it also seems a bit awkward to initialize by PPU device information, in order to then feed it with something else, except for comparison. In other words, I am not sure if a generic class with specific methods would be better than a specific class with a generic method by argument 🙃

Reply:
I'm not sure what you mean here. The line you highlighted here is only to make it convenient to use with `KeyData` or `SourceData`. Otherwise one would need to first filter the `DataCollection` and then make source/keydata object. Here you can instantiate all your object at the start of your script and then filter trains for any object as required.

Reply:
```
ppu = PPU(run)
filtered_run = ppu.trains()
source filtered_run['awesome_source']
```
vs
```
ppu = PPU(run)
source = run['awesome_source']
filtered_source = ppu.trains(source)
```

Reply:
Thanks for the explanation. My comment was not about that particular line in the first place, more about the way this method accepts other source data i.e. allowing to do train picking based on a pre-selected source. This may be more convenient indeed. The default filtering i.e. ``trains()`` method without argument is using the PPU device source, if one is actually interested in another source as in your first code block, it is more "natural" to do it the 2nd way.

Reply:
Still, on a much higher level, I thought whether it would be good have a very generic "trainPicker" or "trainSelector" component in EXtra, that is, a class that could take PPU or any other source for filtering, already at instantiation level.

Reply:
So this comment is not affecting the PR, it is more an outlook to the future. (Maybe it does not make sense, because it lacks a real use case?)

Reply:
Why not, but do we have other filtering sources/methods at this moment?